### PR TITLE
2D-TP & FP8 KV Support for MLA + replicated attention weights

### PR DESCRIPTION
**Pull request by gpolovets1**
# Description

Updated shardings to support MLA for 2D-TP (including adding new `attn_dp_expert` axis)
Also included support for using replicated attention weights and custom compilation batch sizes.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/sharding.py [hunk 1]
```diff
@@ -155,20 +165,25 @@ def from_vllm_config(cls,
...
             # Use attention DP instead to reduce per-device num_kv_heads and
             # eliminate this waste.
-            num_kv_heads_per_device_in_kv_cache = (num_kv_heads * 2) / packing
+            
+            num_kv_heads_per_device_in_kv_cache = max(1, (num_kv_heads * 2) / packing)
             attn_dp = max(
                 int(tensor_parallelism // num_kv_heads_per_device_in_kv_cache),
                 1)
             tensor_parallelism = tensor_parallelism // attn_dp
+            attn_dp_expert = expert_parallelism
```
Comment (wenxindongwork) [inline tpu_inference/layers/common/sharding.py:179]: I'm a bit confused on why we need a new axis `attn_dp_expert` here, could you help me understand why we can't simply use `expert`? would one of `expert` and  `attn_dp_expert` always be one?

Comment (bzgoogle) [reply tpu_inference/layers/common/sharding.py:179]: +1. also curious the different usage between expert and attn_dp_expert ?

Comment (gpolovets1) [reply tpu_inference/layers/common/sharding.py:179]: It is intended to be the analogy to 'model' & 'attn_dp': https://github.com/vllm-project/tpu-inference/pull/1660/changes#diff-de37130aad6bcb0c3aef2904bc90bc0bf03971954abc55ef5371a0eb2903a68eR177

Since we are sharding the model weights now on both 'model' and 'expert', we want to shard the kv cache on both of these axes and create model * expert DP shards. Without this, total_dp_size would be just 'attn_dp' which does not include the shards from the 'expert' axis.

But if there is a cleaner way to do this than applying the analogy of model/attn_dp, ideas are welcome!

Comment (wenxindongwork) [inline tpu_inference/layers/common/sharding.py:179]: In the JAX path, we don't enable expert parallelism via the `--enable_expert_parallelism` flag, but instead do it via `additional_config`, is this understanding correct? (also is there a reason why we haven't unified these approaches?) 

Wondering how this new change works on the torchax path. In the torchax path, we don't use the `expert` axis. We use the `model` axis instead. In that case, there will be no way to use this 2D tensor sharding in torchax right?

Comment (gpolovets1) [reply tpu_inference/layers/common/sharding.py:179]: That's a good question, @jrplatin has been working on unifying the code for DeepSeek. Do we still plan on using the 'expert' axis from additional_config to create the weight shardings?
In general, I have not checked out of the box support on vllm path for DeepSeek (we have been running it through the Jax path) but can add that as a future todo (which may require a few more changes like what you are bringing up with enable_expert_parallelism).

Comment (wenxindongwork) [reply tpu_inference/layers/common/sharding.py:179]: this approach makes sense to me! I can't think of any cleaner approach

Comment (gpolovets1) [reply tpu_inference/layers/common/sharding.py:179]: Aligned async that follow-up PRs will address other sharding configurations that are also used by torchax.

File: tpu_inference/layers/common/quantization/__init__.py [hunk 2]
```diff
@@ -271,11 +271,13 @@ def static_per_tensor_quantize_tensor(
...
-    value: jax.Array,
-    k_scale: float,
-    v_scale: float,
+    value: jax.Array | None = None,
+    k_scale: float | None = None,
+    v_scale: float | None = None,
 ) -> Tuple[jax.Array, jax.Array]:
     """Static quantize key and value tensors."""
     key = static_per_tensor_quantize_tensor(dtype, key, k_scale)
+    if value is None:
```
Comment (jrplatin) [inline tpu_inference/layers/common/quantization/__init__.py:280]: Can you add a comment explaining this case more -- not immediately obvious to me why this is needed, thanks

File: tpu_inference/layers/common/sharding.py [hunk 3]
```diff
@@ -128,14 +134,22 @@ def from_vllm_config(cls,
         sharding_strategy = vllm_config.additional_config.get(
             "sharding", {}).get("sharding_strategy", {})
         parallel_config = vllm_config.parallel_config
-        tensor_parallelism = parallel_config.tensor_parallel_size
+        # Currently tensor_parallelism is also used for other things like determining number of Ray workers.
+        pc_tensor_parallelism = parallel_config.tensor_parallel_size
```
Comment (jrplatin) [inline tpu_inference/layers/common/sharding.py:138]: Nit: maybe write out the full abbreviation? Not obvious what PC/SS are

File: tpu_inference/runner/tpu_runner.py [hunk 4]
```diff
@@ -403,12 +405,22 @@ def _init_inputs(self) -> None:
...
+            "compilation_sizes", [])
         # [16, 32, 64, 128, 256, 512, 1024, 2048]
+        cache_dtype = self.cache_config.cache_dtype
+        if cache_dtype == "auto":
+            cache_dtype = self.dtype
+        kv_cache_dtype = to_jax_dtype(cache_dtype)
+        kv_packing = common_utils.get_dtype_packing(kv_cache_dtype)
         self.num_tokens_paddings = runner_utils.get_token_paddings(
-            min_token_size=max(16, self.dp_size),
+            min_token_size=max(16, self.dp_size * kv_packing),
```
Comment (jrplatin) [inline tpu_inference/runner/tpu_runner.py:418]: Why is this needed?

File: tpu_inference/runner/tpu_runner.py [hunk 5]
```diff
@@ -403,12 +405,22 @@ def _init_inputs(self) -> None:
...
             max_token_size=scheduler_config.max_num_batched_tokens *
             self.dp_size,
             padding_gap=vllm_envs.VLLM_TPU_BUCKET_PADDING_GAP)
+        self.num_tokens_paddings = sorted(self.num_tokens_paddings +
```
Comment (jrplatin) [inline tpu_inference/runner/tpu_runner.py:422]: Why is this needed?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 6]
```diff
@@ -1568,11 +1585,24 @@ def _create_deepseek_attention(
                 query_tnh_spec = P(ShardingAxisName.MLP_TENSOR, None, None)
                 keyvalue_skh_spec = P(ShardingAxisName.MLP_TENSOR, None)
                 attn_o_tnh_spec = P(ShardingAxisName.MLP_TENSOR, None, None)
-
+                anh_sharding = (None, ShardingAxisName.MLP_TENSOR, None)
             else:
                 query_tnh_spec = P(None, ShardingAxisName.MLP_TENSOR, None)
                 keyvalue_skh_spec = P(None, ShardingAxisName.MLP_TENSOR, None)
                 attn_o_tnh_spec = P(None, ShardingAxisName.MLP_TENSOR, None)
+            rd_sharding = (ShardingAxisName.MLP_TENSOR, None)
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:1613]: Are these variables names in the glossary?

File: tpu_inference/layers/common/quantization/__init__.py (hunk 2 above)
Comment (gpolovets1) [reply tpu_inference/layers/common/quantization/__init__.py:280]: Hi @jrplatin it's because for MLA and some other models, there is a latent that is shared between both K & V. In these cases, you only need to quantize a single array instead of two arrays.

File: tpu_inference/runner/tpu_runner.py (hunk 4 above)
Comment (gpolovets1) [reply tpu_inference/runner/tpu_runner.py:418]: IIRC, for multi host (maybe in conjunction with MLA), there was a bug because the dimension was further divided by kv_packing and made the overall dimension size < 1 (i.e. a fraction). This was necessary to fix it.

File: tpu_inference/runner/tpu_runner.py (hunk 5 above)
Comment (gpolovets1) [reply tpu_inference/runner/tpu_runner.py:422]: This is to allow custom compilation shapes that are not necessarily powers of 2. E.g. if your batch size becomes large, then unless it is a power of 2, there could be a very large amount of padding. And your maximum concurrency does not always line up with a power of 2.

File: tpu_inference/layers/common/sharding.py (hunk 3 above)
Comment (gpolovets1) [reply tpu_inference/layers/common/sharding.py:138]: Could you provide an example of what you're thinking? I was initially trying to compromise between length and clarity.

File: tpu_inference/models/jax/deepseek_v3.py (hunk 6 above)
Comment (gpolovets1) [reply tpu_inference/models/jax/deepseek_v3.py:1613]: What do you mean by glossary? They were already previously defined in line 1612 below. I just allow for the option to keep these weights replicated.
